### PR TITLE
add treefmt.withConfig

### PR DESCRIPTION
**Pull request by zimbatm**
This is useful if you want to configure treefmt with nix, and precisely
pass all the commands from nixpkgs.

### CODE REVIEW

**Comment by srid** [comment]
cf https://github.com/numtide/treefmt/issues/165#issuecomment-1124036328

File: .envrc [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export PRJ_ROOT=$PWD
```
Comment (srid) [inline .envrc:1]: Interesting trick! Does `direnv` also use project root as `PWD` even if you open the terminal directly in one of its sub directories?

**Comment by srid** [comment]
@zimbatm What do you think of exposing a `flakeModule` (a la https://github.com/cachix/pre-commit-hooks.nix/pull/130) so that projects that already use `flake-parts` can use this directly, thus resolving #165?

File: .envrc (hunk 1 above)
Comment (zimbatm) [reply .envrc:1]: Yes, this is necessary so that relative paths in the `.envrc` have the same result, even if the user jumps into a sub-folder directly.

**Comment by zimbatm** [comment]
Sounds good, the flake module probably can call treefmt.withConfig with its resulting schema at the end.

File: default.nix [hunk 2]
```diff
@@ -24,13 +28,28 @@ let
     cargoLock.lockFile = ./Cargo.lock;
 
     meta.description = "one CLI to format the code tree";
+
+    passthru.withConfig = { ... }@config:
```
Comment (zimbatm) [inline default.nix:32]: TODO: use the module system and expose the settings as well.

**Comment by srid** [comment]
> Sounds good, the flake module probably can call treefmt.withConfig with its resulting schema at the end.

Great; I'd be happy to test it out here:

https://github.com/srid/haskell-template/blob/6c9e2eb0870b108731df87cdcbb016c064e55041/flake.nix#L19-L25

File: default.nix (hunk 2 above)
Comment (zimbatm) [reply default.nix]: Done

**Comment by zimbatm** [comment]
@srid ok this is ready if you want to give it a try

**Comment by srid** [comment]
@zimbatm Trying it out on haskell-template,

```diff
diff --git a/flake.nix b/flake.nix
index 9918a57..684242e 100644
--- a/flake.nix
+++ b/flake.nix
@@ -5,22 +5,23 @@
     flake-parts.url = "github:hercules-ci/flake-parts";
     flake-parts.inputs.nixpkgs.follows = "nixpkgs";
     haskell-flake.url = "github:srid/haskell-flake";
+    treefmt.url = "github:numtide/treefmt/withConfig";
+    treefmt.flake = false;
   };
 
-  outputs = { self, nixpkgs, flake-parts, haskell-flake, ... }:
+  outputs = inputs@{ self, nixpkgs, flake-parts, haskell-flake, treefmt, ... }:
     flake-parts.lib.mkFlake { inherit self; } {
       systems = nixpkgs.lib.systems.flakeExposed;
       imports = [
         haskell-flake.flakeModule
       ];
-      perSystem = { self', pkgs, ... }: {
+      perSystem = { self', system, pkgs, ... }: {
         haskellProjects.default = {
           root = ./.;
           buildTools = hp: {
-            # TODO: Use https://github.com/numtide/treefmt/pull/169
             inherit (pkgs)
-              treefmt
               nixpkgs-fmt;
+            treefmt = (import treefmt { inherit system; nixpkgs = pkgs; }).treefmt.withConfig (pkgs.lib.importTOML ./treefmt.toml);
             inherit (hp)
               cabal-fmt
               fourmolu;
```

I see:

```
srid on now haskell-template on  master [!] via λ 9.0.2 via ❄️  impure (ghc-shell-for-haskell-template-0.1.0.0) 
❯ more $(which treefmt)
#!/nix/store/14lypyys4gfcl982rjddxa6jg7msqz9q-bash-5.1-p16/bin/bash
exec /nix/store/d88p6rcrpp83ir7i6lasljf871j34g25-treefmt/bin/treefmt --config-file /nix/store/vfblhmhr809df21cvcnh7mdwpqn0nv9w-treefmt.toml "$@"


srid on now haskell-template on  master [!] via λ 9.0.2 via ❄️  impure (ghc-shell-for-haskell-template-0.1.0.0) 
❯ treefmt
[ERR]: If --config-file is set, --tree-root must also be set
```

I wanted to see how `importTOML` behaved before switching to config defined entirely in Nix, but as you can see it looks like I need this PRJ_ROOT env var and pass it via `--tree-root`, which in turn would require direnv or devshell. Therefore, it seems to be that `treefmt.withConfig` as it is implemented currently wouldn't work with projects that do not use either of direnv or devshell.

**Comment by zimbatm** [comment]
agreed. And since the absolute path to the project root would vary from machine to machine, I don't have a really good solution for it. Unless maybe the function also takes a command that points to the project root, like `git rev-parse --show-toplevel`

**Comment by srid** [comment]
What if we had a wrapper script to determine the project root by traversing until a parent directory with a `flake.nix` in it? It does assume that the project has no sub-flakes in it.

**Comment by srid** [comment]
> What if we had a wrapper script to determine the project root by traversing until a parent directory with a `flake.nix` in it? It does assume that the project has no sub-flakes in it.

We can refine this idea to avoid the sub-flake assumption as follows:

```nix
formatter = pkgs.treefmt.withConfig (nixpkgs.lib.importTOML ./treefmt.toml) { projectRootFile = "flake.nix" };
```

Basically, this tells the wrapper script to `cd ../.` until the `projectRootFile` can be located. If the project has sub-flakes, the user can set it to an alternative file, like `{ projectRootFile = "cabal.project"; }`.

@zimbatm If we can get this feature merged, it would be useful in https://github.com/srid/treefmt-flake/issues/1. Ultimately, I imagine that `treemfmt-flake` can be upstreamed to the flake.nix of this repo.

**Comment by zimbatm** [comment]
yeah, that makes sense. Do you have the time to finish this PR so it takes `pkgs.treefmt.withConfig { config = ...; projectRootFile = "..."; }`, all of this gets passed to the module system, and it outputs the wrapper?

**Comment by srid** [comment]
@zimbatm Let me know what you think: https://github.com/numtide/treefmt/pull/177

**Comment by zimbatm** [comment]
ok, I think this is ready now. Any last words/feedback?

File: flake.nix [hunk 3]
```diff
@@ -20,6 +20,14 @@
 
         packages = pkgs;
 
+        # In Nix 2.8 you can run `nix fmt` to format this whole repo. Note that you need to have loaded the
+        # `nix develop` shell before so the various formatters are available in the PATH.
```
Comment (zimbatm) [inline flake.nix:24]: that's true only if the dependencies are not specified in nix. if the config has references to packages then it should work even outside of `nix develop`

**Comment by srid** [comment]
@zimbatm I'm testing this out. I had to apply this diff:

```diff
diff --git a/default.nix b/default.nix
index fc46f6e..d55fff6 100644
--- a/default.nix
+++ b/default.nix
@@ -153,7 +153,7 @@ let
       let
         mod = evalModule config;
       in
-      mod.build.wrapper;
+      mod.config.wrapper;
   };
```

But `nix fmt` still doesn't work:

image

**Comment by srid** [comment]
Also `withConfig` is currently returning a set (`trace: { package = <CODE>; projectRootFile = <CODE>; }`), which forces me to use it like:

```nix
            treefmt = (inputs'.treefmt.packages.default.withConfig {
              settings = nixpkgs.lib.importTOML ./treefmt.toml;
              wrapper.projectRootFile = "flake.nix";
            }).package;
```

(Note the `.package` at the end).

Wouldn't it be simpler to be able to write it without the `.package`?  I.e, have `withConfig` directly return the package.

**Comment by srid** [comment]
> ok, I think this is ready now. Any last words/feedback?

I've tested it to work after applying #178

Looks good to merge. I'd only suggest making `wrapper.projectRootFile = "flake.nix";` the default so the user doesn't have to specify it explicitly on every project using flakes (but without a sub-flake).

**Comment by zimbatm** [comment]
Rebased. It looks like flake.parts doesn't support the formatter option yet. I will give it another shot next week.

File: default.nix [hunk 4]
```diff
@@ -13,6 +13,120 @@ let
...
   cargoToml = with builtins; (fromTOML (readFile ./Cargo.toml));
 
+  # A new kind of option type that calls lib.getExe on derivations
+  exeType = lib.mkOptionType {
+    name = "exe";
+    description = "Path to executable";
+    check = (x: lib.isString x || lib.isPath x || lib.isDerivation x);
+    merge = loc: defs:
+      let res = lib.mergeOneOption loc defs; in
+      if lib.isString res || lib.isPath res then
```
Comment (srid) [inline default.nix:23]: When passing a package, this is failing with:

```
error: attribute 'isPath' missing

       at /nix/store/n7v25mkag2k3g170ljfkppg5aj07hqv5-source/default.nix:20:35:

           19|     description = "Path to executable";
           20|     check = (x: lib.isString x || lib.isPath x || lib.isDerivation x);
             |                                   ^
           21|     merge = loc: defs:
```

Should it be `builtins.isPath` instead? However, `lib.isPath` should work in theory, going by [what's in nixpkgs](https://github.com/NixOS/nixpkgs/search?q=isPath&type=code).

File: default.nix [hunk 5]
```diff
@@ -13,6 +13,120 @@ let
 
   cargoToml = with builtins; (fromTOML (readFile ./Cargo.toml));
 
+  # A new kind of option type that calls lib.getExe on derivations
+  exeType = lib.mkOptionType {
+    name = "exe";
+    description = "Path to executable";
+    check = (x: lib.isString x || lib.isPath x || lib.isDerivation x);
```
[the remark this answers is not in the document]
Comment (zimbatm) [reply default.nix]: yes, turns out lib.isPath is not part of nixpkgs/lib

**Comment by zimbatm** [comment]
Blocked on https://github.com/hercules-ci/flake-parts/pull/56

File: .envrc (hunk 1 above)
Comment (MatthewCroughan) [reply .envrc]: `DIRENV_DIR` already exists and can be queried if you use direnv, there may be no need to set `PRJ_ROOT` then.

**Comment by zimbatm** [comment]
ok, I think this is ready

File: .envrc [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+# Environment variables for the project. See https://direnv.net/
+if nix flake info &>/dev/null; then
```
Comment (Mic92) [inline .envrc:3]:
```suggestion
if nix flake metadata &>/dev/null; then
```

**Comment by srid** [comment]
@zimbatm Will this (`withConfig`) be upstreamed to nixpkgs at some point?

**Comment by zimbatm** [comment]
I'm unsure about the policy regarding nix modules in the pkgs/ tree. Are there any precedents?

**Comment by srid** [comment]
I have seen `passthru` used by certain packages, eg: https://github.com/NixOS/nixpkgs/blob/45b92369d6fafcf9e462789e98fbc735f23b5f64/nixos/modules/services/networking/stubby.nix#L31

**Comment by zimbatm** [comment]
that's not quite the same, plus it's using IFD :-D

With @Mic92 we discussed moving this to another repo, so all the module systems could live separate from the code, and include pre-made samples for the various programs.

**Comment by zimbatm** [comment]
Now live at https://github.com/numtide/treefmt-nix . Looking for feedback!

**Comment by srid** [comment]
> Looking for feedback!

Sure: https://github.com/numtide/treefmt-nix/issues/2

**Pull request merged**